### ISSUE TITLE
Backgrounding app clears reply

### ISSUE THREAD

**Issue by fishcharlie** [damus-io/damus#1862]
## Expected Behavior

When I background the app and reopen it, I expect my draft reply to remain in tact.

## Current Behavior

When I background the app and reopen it, my reply is dismissed and I lose the reply I was working on.

## Steps to reproduce

1. Open Damus
2. Navigate to notifications tab
3. Tap reply button
4. Type reply
5. Multitask to another app
6. Interact with app
7. Multitask back to Damus
8. Observe that reply is discarded

## Video Recording

https://github.com/damus-io/damus/assets/860375/4cfd79f0-e0cf-48a9-b3a3-79b028a86fe2

## Device Information

- iPhone 15 Pro Max
- iOS 17.2.1
- Damus 84cfeb16048c6387a38eaac5d1fbf1a1d1762a13

**Comment by alltheseas** [comment]
I cannot recreate using the steps above

**Comment by alltheseas** [comment]
https://damus.io/nevent1qqsqdhvry050kkk8pds42ztqw7c2jjwxymk5cfpw05fw20jpnawymncpz4mhxue69uhhwmm59eekjctdwd68ytnrdakszrthwden5te0dehhxtnvdakqzxrhwden5te0wahhgtnnw4jx7cmpwfkx7uewvdhk6qg5waehxw309amk7apwva5hy6twduhx7un8snakqy

> @jb55 type the perfect long note on @damus iOS but don’t send it. Switch to Firefox and search for the perfect gif. Save it to photos. Switch back to include the perfect gif. After loading, the draft is gone. Retype the perfect note from flawed memory. Cry.

**Comment by danieldaquino** [comment]
Quick update on this: I am working on making drafts persistent to improve the experience, however it seems that the biggest challenge will be making `DraftArtifacts` conform to `Codable`, given that it (and the whole `PostView` stack) depends on `NSMutableAttributedString` which is not `Codable`.

I see two approaches to get there:
1. Making a specialized data structure for editable nostr post content (Probably using `[Block]` from `Block.swift`), and customize our text editor further to use this as the backing data model. I have done a similar thing once in the past when I created a WYSIWYG algebraic equation editor using `UITextInput` and related TextKit protocols for a personal project, so I am familiar with those protocols. However, it would require a very large refactor around `PostView` and `TextViewWrapper`, so I would estimate this to be between 8–13 story points to complete. It would incidentally give us more control over the text editing experience.
2. Try to encode/decode `NSAttributedString` (or at least the attributes that are used by PostView) for the purpose of making drafts persistent. I estimate this to be about 3 story points to complete.


I started with approach (1), but now I am planning to shift back to (2), as I realized most people seem to be happy with the current text editor, and perhaps such a large and risky refactor is not really worth it at this point. (1) would be a better option only if we are very unhappy with the current text editor and wanted more control over the text editing experience.

@jb55, any thoughts before I move ahead with option (2)?

**Comment by alltheseas** [comment]
possibly same as https://github.com/damus-io/damus/issues/2472

**Comment by jb55** [comment]
I would also prefer (1), but (2) is fine for now. the proper way to do it is with https://github.com/nostr-protocol/nips/blob/master/37.md and nip44, but that is stuff we don't have available yet (although there is some nip44 code in the swift nostr lib I think)

I would imagine in the nip37 case you would store the text in serialized block form and then you would simply parse it again when editing, rebuilding the editor view from blocks.

**Comment by danieldaquino** [comment]
@jb55, @alltheseas, a quick status update on this:

**These are the things that are currently done:**
- Drafts can be converted into NIP-37 draft notes, and vice-versa
- Those NIP-37 drafts are being saved to NostrDB persistently, and successfully loaded on app start
    - I am intentionally NOT saving them to relays for now because that will require more complex sync-down logic, it has some privacy implications, and that is out of the scope of this ticket.
- Draft auto-saves while user works through the note (periodically, to avoid performance issues)
- Draft loads rich info (i.e. it loads mentions with the purple link, the user name, attached images load as expected, etc)
- All of the above are working for the different kinds of drafts (Regular post, replies, quotes, highlights... They can all save separate drafts)

**What's left:**
- There are still some issues/weirdness with parsing between NIP-37 drafts and draft artifacts (e.g. Quoted notes end up being quoted twice the second time it is loaded). In other words, it's not fully robust yet.
- Code cleanup, documentation, and validation testing

Basically, the functionality is there and mostly working, but some refinement is still needed.

**Comment by danieldaquino** [comment]
@jb55 @alltheseas sent the patch for this one with all details via email to patches@damus.io. Please let me know if you have any suggestions.

Thanks!
